### DISCUSSION TITLE
422 Unprocessable Entity with TestClient

### DISCUSSION THREAD

**Discussion by hgue** [fastapi/fastapi#5907] [Questions] [answered]
Hello,
sorry for asking this here, but I am totally out of ideas with my case regarding testing the fastapi api using the TestClient. Hope someone can lead me in the right direction.

Please look at the following code:

####################################################################

class CoreModel(BaseModel):
    """
    Any common logic to be shared by all models goes here
    """
    pass

class UserCreate(CoreModel):
    """
    Email, username, and password are required for registering a new user
    """
    email: EmailStr
    password: constr(min_length=7, max_length=100)
    username: constr(min_length=3, regex="[a-zA-Z0-9_-]+$")

    class Config:
        # The lineorm_mode = True allows the app to take ORM objects and translate
        # them into responses automatically. This automation saves us from manually
        # taking data out of ORM, making it into a dictionary, then loading it in with Pydantic.
        orm_mode = True

@router.post("/", status_code=status.HTTP_201_CREATED, name="users:register-new-user")
def register_new_user(
    new_user: UserCreate = Body(..., embed=True),
    users_service: UsersService = Depends(get_service(UsersService, UsersRepository))
) -> UserPublic:

    result = users_service.register_new_user(new_user)
    created_user = result.handle()
    ...

@pytest.fixture
def user_to_create():
    return {"new_user": {"email": "shakira@shakira.io", "password": "chantaje", "username": "shakirashakira"}}
    
def test_users_can_register_successfully(client: TestClient, user_to_create):
    users_service = UsersService(client.db, UsersRepository)

    res = client.post(client.app.url_path_for("users:register-new-user"), json=user_to_create)
        
    assert res.status_code == status.HTTP_201_CREATED

####################################################################

The last line thows the following exception when using pytest an 

2023-01-21 09:46:27 | httpx._client | _client | HTTP Request: POST http://testserver/api/v1/users/ "HTTP/1.1 422 Unprocessable Entity"
{'new_user': {'email': 'shakira@shakira.io', 'password': 'chantaje', 'username': 'shakirashakira'}}
{'detail': [{'loc': ['query', 'kw'], 'msg': 'field required', 'type': 'value_error.missing'}]}
FAILED

Somehow it is not process the 'new_user' object correctly. I am very sure, the test data is correct, it is working with the swagger documentation on the website using the endpoint with the same dataset. So maybe there is something missing or a problem with TestClient class?
I tried a lot, but have no idea howto solve this...

**Comment by hgue** [comment]
Hello,
I made a public repository here: [unprocessable_entity](https://gitlab.com/webtux/unprocessable_entity)

1. Install [poetry package manager](https://python-poetry.org)
2. In folder backend run `poetry shell`
3. run `pytest -s -v app/core/tests/api/test_users.py`

**Reply by chris14jan** [reply]
The [unprocessable_entity](https://gitlab.com/webtux/unprocessable_entity) gitlab link is broken

**Comment by RonRothman** [comment]
My guess (based on something similar I just debugged) is that it's related to these overrides you're making:

```
        app.dependency_overrides[get_session] = db_session
        app.dependency_overrides[get_service] = get_serv
```

To confirm, try settings those overrides to a simple class and see what happens.

**Comment by dkimbel** [comment, marked as answer]
I had a very similar issue recently. My problem was that the function I assigned to `app.dependency_overrides` was expecting one argument too many -- it was expecting one argument when it should have been expecting zero.

My code was like this:
```
app.dependency_overrides[some_dependency] = lambda _: some_return_value
```

And it needed to be this:
```
app.dependency_overrides[some_dependency] = lambda: some_return_value
```

In my case, instead of `'loc': ['query', 'kw']`, my error message said `'loc': ['query', '_']`  -- where `_` was the literal name of my extra argument.

**Reply by RafaelWO** [reply]
I was debugging my test for hours... I made the same mistake of creating an override via `lambda _: some_return_value
` and got `{"detail":[{"type":"missing","loc":["query","_"],"msg":"Field required","input":null}]}` as my JSON body in the HTTP 422 response. Thanks for your comment, @dkimbel!

**Comment by YuriiMotov** [comment]
Link to the repo from initial question doesn't work, but I believe the problem is the same that @dkimbel
 described here: https://github.com/fastapi/fastapi/discussions/5907#discussioncomment-5767938
 
Closing this
